plugin-server/src/worker/vm/upgrades/utils/export-events-buffer.ts

```diff
@@ -0,0 +1,68 @@
+import { Hub } from 'types'
+
+export type BufferOptions = {
+    limit: number
+    timeoutSeconds: number
+    onFlush?: (objects: any[], points: number) => void | Promise<void>
+}
+
+export class ExportEventsBuffer {
+    buffer: any[]
+    timeout: NodeJS.Timeout | null
+    points: number
+    options: BufferOptions
+    hub: Hub
+
+    constructor(hub: Hub, opts?: Partial<BufferOptions>) {
+        this.buffer = []
+        this.timeout = null
+        this.points = 0
+        this.options = {
+            limit: 10,
+            timeoutSeconds: 60,
+            ...opts,
+        }
+        this.hub = hub
+    }
+
+    public async add(object: any, points = 1): Promise<void> {
+        // flush existing if adding would make us go over the limit
+        if (this.points && this.points + points > this.options.limit) {
+            await this.flush()
+        }
+
+        // add the object to the buffer
+        this.points += points
+        this.buffer.push(object)
+
+        if (this.points > this.options.limit) {
+            // flush (again?) if we are now over the limit
+            await this.flush()
+        } else if (!this.timeout) {
+            // if not, make sure there's a flush timeout
+            this.timeout = setTimeout(async () => await this.flush(), this.options.timeoutSeconds * 1000)
+        }
+    }
+
+    public async flush(): Promise<void> {
+        this.hub.statsd?.increment(`buffer_voided_promises`, { instanceId: this.hub.instanceId.toString() })
+
+        const oldBuffer = this.buffer
+        const oldPoints = this.points
+        this.buffer = []
+        this.points = 0
+
+        void this.hub.promiseManager.trackPromise(this._flush(oldBuffer, oldPoints, new Date()))
```

Comment:
This `trackPromise` doesn't *really* return a promise. It could as easily just return `void`, and not require an explicit `void` here.

Since we don't (yet) need this function to return the promise, I suggest removing the `return promise` part. We can make a new `wrapPromise` function if / when it's time to implement the babel expression catching code.

Reply:
Good point